### PR TITLE
fix: Unable to import backup - WPB-9372

### PR DESCRIPTION
**Pull request by KaterinaWire**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-9372" title="WPB-9372" target="_blank">BugWPB-9372</a>  [iOS] Beta unable to import backup from Production
  </td></table>
  <br />
 

### Issue
If a user who already has 7 active clients wants to restore a backup created from an app version with an old database version, the app will stuck on the landing screen.

### Solution
After migration, create a new `AuthenticationCoordinator` and go to the `.createCredentials` authentication step. Since we are creating a new authentication coordinator, we no longer have the password from the first step and need to prompt for a password to remove one of the clients. 

**Note:** we didn't ask for a password to delete clients during login, but it wasn't consistent with other platforms. QA and product agree to ask a password each time.

### Testing
1. Create a backup with the previous version of the database.
2. You should have 7 clients for the user you are going to use.
3. Import backup.

**Expected:**
- There should be a screen "Too many devices".
- After deleting one of the clients, the login process should be completed.

### Checklist

- [x] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [x] Description is filled and free of optional paragraphs.
- [x] Adds/updates automated tests.

### CODE REVIEW

**Review by johnxnguyen — APPROVED** [review]
Very nice! I'm glad we found a nice solution.

File: wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/RemoveClients/RemoveClientsViewModel.swift [hunk 1]
```diff
@@ -48,25 +45,16 @@ extension RemoveClientsViewController {
...
-}
 
-private extension ZMEmailCredentials {
-    var emailCredentials: EmailCredentials? {
-        guard let email,
-              let password
-        else {
-            return nil
+            try await removeUserClientUseCase?.invoke(clientId: clientId,
+                                                      credentials: EmailCredentials(email: "", password: password))
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/RemoveClients/RemoveClientsViewModel.swift]: a bit weird that the email is not used, does it make sense to change the interface to accept the password instead?

File: wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/RemoveClients/RemoveClientsViewController.swift [hunk 2]
```diff
@@ -124,16 +138,24 @@ final class RemoveClientsViewController: UIViewController,
...
+
         isLoadingViewVisible = true
         do {
-            try await viewModel.removeUserClient(userClient)
-            isLoadingViewVisible = false
+            try await viewModel.removeUserClient(userClientToDelete, password: password)
             delegate?.finishedDeleting(self)
         } catch {
-            isLoadingViewVisible = false
             delegate?.failedToDeleteClients(error)
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/RemoveClients/RemoveClientsViewController.swift]: Shouldn't we clean the userClientToDelete at some point? I know we teardown everything but just to be safe

**Review by agisilaos — APPROVED** [review]

File: wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/RemoveClients/RemoveClientsViewController.swift [hunk 3]
```diff
@@ -124,16 +138,24 @@ final class RemoveClientsViewController: UIViewController,
         self.navigationController?.presentingViewController?.dismiss(animated: true, completion: nil)
     }
 
-    func removeUserClient(_ userClient: UserClient) async {
+    func removeUserClient(_ userClient: UserClient) {
+        userClientToDelete = userClient
+        present(requestPasswordController.alertController, animated: true)
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/RemoveClients/RemoveClientsViewController.swift]: would it work to have:

```
if let password = await presentRequestPasswordController {
 await removeUserClient(password: password)
}
```

**Review by netbe — COMMENTED** [review]
add some suggestions

File: wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/RemoveClients/RemoveClientsViewModel.swift (hunk 1 above)
Comment (KaterinaWire) [reply wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/RemoveClients/RemoveClientsViewModel.swift]: changed

File: wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/RemoveClients/RemoveClientsViewController.swift (hunk 3 above)
Comment (KaterinaWire) [reply wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/RemoveClients/RemoveClientsViewController.swift]: added

File: wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/RemoveClients/RemoveClientsViewController.swift [hunk 4]
```diff
@@ -138,19 +122,33 @@ final class RemoveClientsViewController: UIViewController,
...
+        }
     }
 
     // MARK: - Helpers
 
-    private func removeUserClient(password: String) async {
-        guard let userClientToDelete = userClientToDelete else { return }
+    private func presentRequestPasswordController() async -> String? {
+        await withCheckedContinuation { continuation in
+            requestPasswordController = RequestPasswordController(
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/RemoveClients/RemoveClientsViewController.swift:135]: do we need to retain RequestPasswordController ? as it is retained by the presentingViewController no?

File: wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/RemoveClients/RemoveClientsViewController.swift [hunk 5]
```diff
@@ -138,19 +122,33 @@ final class RemoveClientsViewController: UIViewController,
...
+                context: .removeDevice,
+                callback: { password in
+                    continuation.resume(returning: password)
+                })
+            guard let alertController = requestPasswordController?.alertController else {
+                return
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/RemoveClients/RemoveClientsViewController.swift:142]: in this case you would need to call continuation in order not to block the execution

Comment (KaterinaWire) [reply wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/RemoveClients/RemoveClientsViewController.swift:142]: done

**Review by netbe — APPROVED** [review]

**Pull request merged**